neurodsp/sim/params.py

```diff
@@ -0,0 +1,589 @@
+
+        Parameters
+        ----------
+        n_seconds : float, optional
+            Simulation time, in seconds.
+        fs : float, optional
+            Sampling rate of simulated signal, in Hz.
+
+        Notes
+        -----
+        If set as False, value will not be updated.
+        If set as None, value will be updated to None.
+        """
+
+        if n_seconds is not False:
+            self.n_seconds = n_seconds
+        if fs is not False:
+            self.fs = fs
+
+
+    def update_param(self, label, attribute, new_value):
+        """Update a simulation parameter definition.
+
+        Parameters
+        ----------
+        label : str
+            Label for the parameter definition.
+        attribute : str
+            Name of the attribute to update.
+        new_value : obj
+            New value to give the attribute.
+        """
+
+        self._params[label][attribute] = new_value
+
+
+    def clear(self, clear_base=False):
+        """"Clear parameter definitions.
+
+        Parameters
+        ----------
+        clear_base : bool, optional, default: False
+            Whether to also clear base parameters.
+        """
+
+        self._params = {}
+
+        if clear_base:
+            self.update_base(None, None)
+
+
+    def _make_params(self, parameters=None, **kwargs):
+        """Sub-function for `make_params`."""
+
+        parameters = {} if not parameters else deepcopy(parameters)
+
+        if isinstance(parameters, list):
+            comps = [parameters.pop(0)]
+            kwargs = {**kwargs, **parameters[0]} if parameters else kwargs
+            params = self._make_combined_params(comps, **kwargs)
+        else:
+            params = {**parameters, **kwargs}
+
+        # If any base parameters were passed in, clear them
+        for bparam in self.base.keys():
+            params.pop(bparam, None)
+
+        return params
+
+
+    def _make_combined_params(self, components, component_variances=None):
+        """Make parameters for combined simulations, specifying multiple components.
+
+        Parameters
+        ----------
+        components : list of dict
+            List of simulation component parameters.
+        component_variances : list of float
+            Component variances for the combined simulation.
+
+        Returns
+        -------
+        params : dict
+            Parameter definition.
+        """
+
+        parameters = {}
+
+        comps = {}
+        for comp in components:
+            comps.update(**deepcopy(comp))
+        parameters['components'] = comps
+
+        if component_variances:
+            parameters['component_variances'] = component_variances
+
+        return parameters
+
+
+class SimIters(SimParams):
+    """Object for managing simulation iterators.
+
+    Parameters
+    ----------
+    n_seconds : float
+        Simulation time, in seconds.
+    fs : float
+        Sampling rate of simulated signal, in Hz.
+    sim_params : SimParams
```

Comment:
It feels odd that this class both inherits from SimParams and accepts a SimParams as initialization argument. It may be worth considering to choose one way or the other if possible.

A hack may be to add a method to SimParams that returns an initialized SimIters object. And then drop sim_params arg here.

Reply:
Yes - agreed! Updated (notes in next comment)